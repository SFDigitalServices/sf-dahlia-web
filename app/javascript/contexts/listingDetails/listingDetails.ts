interface ListingDetails {
  /**  */
  uid: string

  /**  */
  email: string

  /**  */
  created_at: Date

  /**  */
  updated_at: Date
}

interface ListingDetailsData {
  data: any
  success: boolean
}

export { ListingDetails, ListingDetailsData }
