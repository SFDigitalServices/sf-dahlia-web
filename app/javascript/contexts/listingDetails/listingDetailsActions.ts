import { createAction } from "typesafe-actions"
import { RailsUnit } from "../../api/types/rails/listings/RailsUnit"
import { RailsAmiChart } from "../../api/types/rails/listings/RailsAmiChart"

export enum ListingDetailsActions {
  StartFetchingUnits = "SetFetchingUnits",
  FinishFetchingUnits = "FinishFetchingUnits",
  SetUnits = "SetUnits",
  SetAmiCharts = "SetAmiCharts",
  StartFetchingAmiCharts = "SetFetchingAmiCharts",
  SetFetchedAmiCharts = "SetFetchedAmiCharts",
  SetFetchingAmiChartsError = "SetFetchingAmiChartsError",
  SetFetchingUnitsError = "SetFetchingUnitsError",
  FinishFetchingAmiCharts = "FinishFetchingAmiCharts",
}

export const startFetchingUnits = createAction(ListingDetailsActions.StartFetchingUnits)()
export const setUnits = createAction(ListingDetailsActions.SetUnits)()
export const setAmiCharts = createAction(ListingDetailsActions.SetAmiCharts)()
export const finishFetchingAmiCharts = createAction(ListingDetailsActions.FinishFetchingAmiCharts)<
  RailsAmiChart[]
>()
export const setFetchedAmiCharts = createAction(ListingDetailsActions.SetFetchedAmiCharts)()
export const startFetchingAmiCharts = createAction(ListingDetailsActions.StartFetchingAmiCharts)()
export const finishFetchingUnits = createAction(ListingDetailsActions.FinishFetchingUnits)<
  RailsUnit[]
>()
export const setFetchingUnitsError = createAction(
  ListingDetailsActions.SetFetchingUnitsError
)<Error>()
export const setFetchingAmiChartsError = createAction(
  ListingDetailsActions.SetFetchingAmiChartsError
)<Error>()
