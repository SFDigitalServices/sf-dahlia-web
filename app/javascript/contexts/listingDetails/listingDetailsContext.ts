import { createContext } from "react"
import RailsUnit from "../../api/types/rails/listings/RailsUnit"
import { RailsAmiChart, RailsAmiChartMetaData } from "../../api/types/rails/listings/RailsAmiChart"

export type ContextProps = {
  units: RailsUnit[]
  amiCharts: RailsAmiChart[]
  fetchedAmiCharts: boolean
  fetchingAmiCharts: boolean
  fetchedUnits: boolean
  fetchingUnits: boolean
  fetchUnits: (listingId: string) => void
  fetchAmiCharts: (chartsToFetch: RailsAmiChartMetaData[]) => void
  fetchingAmiChartsError: Error
  fetchingUnitsError: Error
}

export default createContext<Partial<ContextProps>>({})
