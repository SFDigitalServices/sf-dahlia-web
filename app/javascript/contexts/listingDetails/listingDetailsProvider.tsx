import React, { useReducer } from "react"
import ListingDetailsReducer from "./listingDetailsReducer"
import ListingDetailsContext, { ContextProps } from "./listingDetailsContext"
import { getUnits, getAmiCharts } from "../../api/listingApiService"
import { RailsUnit } from "../../api/types/rails/listings/RailsUnit"
import { RailsAmiChart, RailsAmiChartMetaData } from "../../api/types/rails/listings/RailsAmiChart"
import {
  finishFetchingUnits,
  finishFetchingAmiCharts,
  startFetchingAmiCharts,
  startFetchingUnits,
  setFetchingAmiChartsError,
  setFetchingUnitsError,
} from "./listingDetailsActions"

interface ListingDetailsProviderProps {
  children?: React.ReactNode
}

const ListingDetailsProvider = (props: ListingDetailsProviderProps) => {
  const [state, dispatch] = useReducer(ListingDetailsReducer, {
    units: [],
    amiCharts: [],
    fetchingUnits: false,
    fetchedUnits: false,
    fetchingAmiCharts: false,
    fetchedAmiCharts: false,
    fetchingAmiChartsError: null,
    fetchingUnitsError: null,
  })

  const contextValues: ContextProps = {
    units: state.units,
    amiCharts: state.amiCharts,
    fetchedAmiCharts: state.fetchedAmiCharts,
    fetchingAmiCharts: state.fetchingAmiCharts,
    fetchedUnits: state.fetchedUnits,
    fetchingUnits: state.fetchingUnits,
    fetchingAmiChartsError: state.fetchingAmiChartsError,
    fetchingUnitsError: state.fetchingUnitsError,
    fetchUnits: (listingId: string) => {
      dispatch(startFetchingUnits())
      getUnits(listingId)
        .then((units: RailsUnit[]) => {
          dispatch(finishFetchingUnits(units))
        })
        .catch((error) => {
          console.error(error)
          dispatch(setFetchingUnitsError(error))
        })
    },
    fetchAmiCharts: (chartsToFetch: RailsAmiChartMetaData[]) => {
      dispatch(startFetchingAmiCharts())
      getAmiCharts(chartsToFetch)
        .then((amiCharts: RailsAmiChart[]) => {
          dispatch(finishFetchingAmiCharts(amiCharts))
        })
        .catch((error) => {
          console.error(error)
          dispatch(setFetchingAmiChartsError(error))
        })
    },
  }

  return (
    <ListingDetailsContext.Provider value={contextValues}>
      {props.children}
    </ListingDetailsContext.Provider>
  )
}

export default ListingDetailsProvider
