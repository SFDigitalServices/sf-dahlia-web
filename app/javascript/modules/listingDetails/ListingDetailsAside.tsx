import React from "react"
import { RailsListing } from "../listings/SharedHelpers"
import { ListingDetailsInfoSession } from "../listingDetailsAside/ListingDetailsInfoSession"
import { ListingDetailsApply } from "../listingDetailsAside/ListingDetailsApply"
import { ListingDetailsProcess } from "../listingDetailsAside/ListingDetailsProcess"
import { isOpen } from "../../util/listingUtil"
import { ListingDetailsApplicationDate } from "../listingDetailsAside/ListingDetailsApplicationDate"

export interface ListingDetailsSidebarProps {
  listing: RailsListing
}

export const ListingDetailsAside = ({ listing }: ListingDetailsSidebarProps) => {
  const isApplicationOpen = isOpen(listing)

  return (
    <aside className="w-full static md:absolute md:right-0 md:w-1/3 md:top-0 sm:w-2/3 md:ml-2 h-full md:border border-solid bg-white">
      <div className="hidden md:block">
        <ListingDetailsApplicationDate listing={listing} />
        {isApplicationOpen && <ListingDetailsInfoSession listing={listing} />}
        <ListingDetailsApply listing={listing} />
        <ListingDetailsProcess listing={listing} />
      </div>
    </aside>
  )
}
