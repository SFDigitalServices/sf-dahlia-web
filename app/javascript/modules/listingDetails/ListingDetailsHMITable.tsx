import React, { useContext, useState, useMemo, useEffect } from "react"
import { Icon, StandardTable, Button, t } from "@bloom-housing/ui-components"
import { RailsListing } from "../listings/SharedHelpers"
import { getMinMaxOccupancy, isSale } from "../../util/listingUtil"
import ListingDetailsContext from "../../contexts/listingDetails/listingDetailsContext"
import { faAngleDown, faAngleUp } from "@fortawesome/free-solid-svg-icons"
import { RailsAmiChart, RailsAmiChartValue } from "../../api/types/rails/listings/RailsAmiChart"
import "./ListingDetailsEligibility.scss"

export interface ListingDetailsEligibilityProps {
  listing: RailsListing
}

const toShowButton = (maxOccupancy: number, minOccupancy: number, hmiCutoff: number): boolean => {
  if (!maxOccupancy) {
    return false
  }

  const calculatedMaxOccupancy = maxOccupancy + 2 - minOccupancy

  if (calculatedMaxOccupancy > hmiCutoff) {
    return true
  }
  return false
}

const buildHmiHeadersWithOneAmi = () => {
  return {
    householdSize: "t.householdSize",
    maxIncomeMonth: "t.maximumIncomeMonth",
    maxIncomeYear: "t.maximumIncomeYear",
  }
}

const hasMultipleUniqueChartsByPercent = (amiCharts: RailsAmiChart[]) => {
  return (
    amiCharts
      .map((chart) => chart.percent)
      .filter((currentPercent, idx, allPercents) => allPercents.indexOf(currentPercent) === idx)
      .length > 1
  )
}

const buildHmiHeadersWithMultipleAmis = (amiCharts: RailsAmiChart[]) => {
  const headers = {
    householdSize: "t.householdSize",
  }
  amiCharts?.forEach((chart: RailsAmiChart) => {
    headers[`ami${chart.percent}`] = `t.percentAMI*percent:${chart.percent}`
  })
  return headers
}

const buildHmiChartHeaders = (amiCharts: RailsAmiChart[]) => {
  return hasMultipleUniqueChartsByPercent(amiCharts)
    ? buildHmiHeadersWithMultipleAmis(amiCharts)
    : buildHmiHeadersWithOneAmi()
}

const sortAmisByPercent = (a: RailsAmiChart, b: RailsAmiChart) => {
  const numA = Number(a.percent)
  const numB = Number(b.percent)

  if (numA < numB) {
    return -1
  }
  if (numA > numB) {
    return 1
  }
  return 0
}

const buildHmiTableWithOneAmiChart = (
  minOccupancy: number,
  maxOccupancy: number,
  amiCharts: RailsAmiChart[]
) => {
  const tableData = []
  const mostRecentAmiChart = amiCharts[amiCharts.length - 1]
  for (let i = minOccupancy; i <= maxOccupancy; i++) {
    const amiChart = mostRecentAmiChart?.values?.find((amiChart: RailsAmiChartValue) => {
      return amiChart.numOfHousehold === i
    })

    const householdSize =
      i === 1
        ? {
            content: <span className="font-semibold">{t("listings.onePerson")}</span>,
          }
        : {
            content: <span className="font-semibold">{`${i} ${t("listings.people")}`}</span>,
          }
    if (amiChart) {
      tableData.push({
        householdSize,
        maxIncomeMonth: {
          content: t("t.perMonthCost", {
            cost: `$${Math.floor(amiChart?.amount / 12).toLocaleString()}`,
          }),
        },
        maxIncomeYear: {
          content: t("t.perYearCost", { cost: `$${amiChart?.amount?.toLocaleString()}` }),
        },
      })
    }
  }

  return tableData
}

const buildHmiTableWithMultipleAmis = (
  minOccupancy: number,
  maxOccupancy: number,
  amiCharts: RailsAmiChart[]
) => {
  const tableData = []
  for (let i = minOccupancy; i <= maxOccupancy; i++) {
    const tableRow = {}

    const householdSize =
      i === 1
        ? {
            content: <span className="font-semibold">{t("listings.onePerson")}</span>,
          }
        : {
            content: <span className="font-semibold">{`${i} ${t("listings.people")}`}</span>,
          }

    amiCharts?.forEach((chart) => {
      const amiChart = chart.values?.find((amiChart) => {
        return amiChart.numOfHousehold === i
      })
      if (amiChart) {
        // if there are charts with the same percent, the chart with the largest index is used
        tableRow[`ami${chart.percent}`] = {
          content: t("t.perYearCost", { cost: `$${amiChart?.amount?.toLocaleString()}` }),
        }
      }
    })

    tableData.push({
      householdSize,
      ...tableRow,
    })
  }

  return tableData
}

const buildHmiCharts = (
  listingIsSale: boolean,
  amiCharts: RailsAmiChart[],
  minOccupancy: number,
  maxOccupancy: number
) => {
  if (!listingIsSale) {
    maxOccupancy += 2
  }

  return hasMultipleUniqueChartsByPercent(amiCharts)
    ? buildHmiTableWithMultipleAmis(minOccupancy, maxOccupancy, amiCharts)
    : buildHmiTableWithOneAmiChart(minOccupancy, maxOccupancy, amiCharts)
}

export interface ReducedUnit {
  name: string
  numberOfUnits: number
}

export const ListingDetailsHMITable = ({ listing }: ListingDetailsEligibilityProps) => {
  const listingIsSale = isSale(listing)
  const {
    fetchingAmiCharts,
    fetchedAmiCharts,
    amiCharts,
    fetchingUnits,
    fetchedUnits,
    units,
    fetchingAmiChartsError,
    fetchingUnitsError,
  } = useContext(ListingDetailsContext)

  useEffect(() => {
    if (fetchingAmiChartsError) {
      // TODO: Log error properly
      throw fetchingAmiChartsError
    }
    if (fetchingUnitsError) {
      // TODO: Log error properly
      throw fetchingUnitsError
    }
  }, [fetchingAmiChartsError, fetchingUnitsError])

  if (amiCharts?.length > 0) {
    amiCharts.sort(sortAmisByPercent)
  }

  const [tableCollapsed, setTableCollapsed] = useState(true)

  const { minOccupancy, maxOccupancy, explicitMaxOccupancy } = useMemo(() => {
    if (!fetchedAmiCharts || !fetchedUnits) {
      return { minOccupancy: undefined, maxOccupnacy: undefined }
    }

    return getMinMaxOccupancy(units, amiCharts)
  }, [units, amiCharts, fetchedAmiCharts, fetchedUnits])

  const HMITableData = useMemo(() => {
    if (!fetchedAmiCharts || !fetchedUnits) {
      return []
    }

    return buildHmiCharts(listingIsSale, amiCharts, minOccupancy, maxOccupancy)
  }, [listingIsSale, amiCharts, fetchedAmiCharts, fetchedUnits, minOccupancy, maxOccupancy])

  const hmiCutoff = explicitMaxOccupancy ? Math.max(Math.floor(maxOccupancy / 2) * 2, 1) : 2

  const HMITableHeaders = useMemo(() => {
    if (!fetchedAmiCharts) {
      return {}
    }
    return buildHmiChartHeaders(amiCharts)
  }, [fetchedAmiCharts, amiCharts])

  if (fetchingUnits || fetchingAmiCharts) {
    return (
      <div className="flex w-full justify-center">
        <Icon symbol="spinner" size="large" />
      </div>
    )
  }

  const expandTableHandler = (): void => {
    setTableCollapsed(!tableCollapsed)
  }

  return (
    <>
      <StandardTable
        headers={HMITableHeaders}
        data={tableCollapsed ? HMITableData.slice(0, hmiCutoff) : HMITableData}
      />
      {toShowButton(maxOccupancy, minOccupancy, hmiCutoff) && (
        <Button
          inlineIcon="right"
          className="font-medium md:bg-primary-lighter mt-3 text-primary-dark"
          iconSize="small"
          icon={tableCollapsed ? faAngleDown : faAngleUp}
          onClick={expandTableHandler}
          ariaLabel={t("listings.householdMaximumIncome.showMore.aria")}
          ariaExpanded={!tableCollapsed}
        >
          {tableCollapsed ? t("label.showMore") : t("label.showLess")}
        </Button>
      )}
    </>
  )
}
