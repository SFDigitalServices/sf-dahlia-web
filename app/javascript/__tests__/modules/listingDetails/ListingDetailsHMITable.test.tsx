import React from "react"
import renderer from "react-test-renderer"
import { ListingDetailsHMITable } from "../../../modules/listingDetails/ListingDetailsHMITable"
import { closedRentalListing } from "../../data/RailsRentalListing/listing-rental-closed"
import { habitatListing } from "../../data/RailsSaleListing/listing-sale-habitat"
import { openSaleListing } from "../../data/RailsSaleListing/listing-sale-open"
import { units } from "../../data/RailsListingUnits/listing-units"
import { amiCharts } from "../../data/RailsAmiCharts/ami-charts"
import ListingDetailsContext from "../../../contexts/listingDetails/listingDetailsContext"

describe("ListingDetailsHMITable", () => {
  it("renders ListingDetailsHMITable component with spinner before api call", () => {
    const tree = renderer.create(
      <ListingDetailsContext.Provider
        value={{
          units: [],
          amiCharts: [],
          fetchingUnits: true,
          fetchedUnits: false,
          fetchingAmiCharts: true,
          fetchedAmiCharts: false,
          fetchingAmiChartsError: null,
          fetchingUnitsError: null,
        }}
      >
        <ListingDetailsHMITable listing={closedRentalListing} />
      </ListingDetailsContext.Provider>
    )
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it("renders ListingDetailsHMITable component with rental listing", () => {
    const tree = renderer.create(
      <ListingDetailsContext.Provider
        value={{
          units,
          amiCharts,
          fetchingUnits: false,
          fetchedUnits: true,
          fetchingAmiCharts: false,
          fetchedAmiCharts: true,
          fetchingAmiChartsError: null,
          fetchingUnitsError: null,
        }}
      >
        <ListingDetailsHMITable listing={closedRentalListing} />
      </ListingDetailsContext.Provider>
    )
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it("renders ListingDetailsPricingTable component with open sale listing", () => {
    const tree = renderer.create(
      <ListingDetailsContext.Provider
        value={{
          units,
          amiCharts,
          fetchingUnits: false,
          fetchedUnits: true,
          fetchingAmiCharts: false,
          fetchedAmiCharts: true,
          fetchingAmiChartsError: null,
          fetchingUnitsError: null,
        }}
      >
        <ListingDetailsHMITable listing={openSaleListing} />
      </ListingDetailsContext.Provider>
    )
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it("does not render ListingDetailsPricingTable when habitat listing", () => {
    const tree = renderer.create(
      <ListingDetailsContext.Provider
        value={{
          units,
          amiCharts,
          fetchingUnits: false,
          fetchedUnits: true,
          fetchingAmiCharts: false,
          fetchedAmiCharts: true,
          fetchingAmiChartsError: null,
          fetchingUnitsError: null,
        }}
      >
        <ListingDetailsHMITable listing={habitatListing} />
      </ListingDetailsContext.Provider>
    )
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
