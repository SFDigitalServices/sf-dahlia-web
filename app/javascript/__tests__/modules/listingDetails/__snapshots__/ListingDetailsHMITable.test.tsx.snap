// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingDetailsHMITable does not render ListingDetailsPricingTable when habitat listing 1`] = `
<DocumentFragment>
  <div
    style="overflow-x: auto;"
  >
    <table
      class="w-full text-xs"
    >
      <thead>
        <tr>
          <th
            class=""
          >
            <span
              class=""
            >
              household size
            </span>
          </th>
          <th
            class=""
          >
            <span
              class=""
            >
              55% AMI (Area Median Income)
            </span>
          </th>
          <th
            class=""
          >
            <span
              class=""
            >
              82% AMI (Area Median Income)
            </span>
          </th>
          <th
            class=""
          >
            <span
              class=""
            >
              109.8% AMI (Area Median Income)
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class=""
          id="standardrow-0"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              1 person
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $51,300 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $61,300 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $71,300 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-1"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              2 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $58,600 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $68,600 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $78,600 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-2"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              3 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $65,950 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $75,950 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $75,950 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-3"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              4 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $73,250 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $83,250 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $83,250 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-4"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              5 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $79,100 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $99,100 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $140,000 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-5"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              6 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $85,000 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $95,000 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $150,000 per year
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <button
    aria-expanded="false"
    aria-label="Show more about maximum income based on household size"
    class="button has-icon-right is-inline font-medium md:bg-primary-lighter mt-3 text-primary-dark"
    tabindex="0"
  >
    <span
      class="button__content"
    >
      Show more
    </span>
    <span
      class="ui-icon ui-small button__icon "
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-angle-down "
        data-icon="angle-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M201.4 342.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 274.7 86.6 137.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
          fill="currentColor"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`ListingDetailsHMITable renders ListingDetailsHMITable component with rental listing 1`] = `
<DocumentFragment>
  <div
    style="overflow-x: auto;"
  >
    <table
      class="w-full text-xs"
    >
      <thead>
        <tr>
          <th
            class=""
          >
            <span
              class=""
            >
              household size
            </span>
          </th>
          <th
            class=""
          >
            <span
              class=""
            >
              55% AMI (Area Median Income)
            </span>
          </th>
          <th
            class=""
          >
            <span
              class=""
            >
              82% AMI (Area Median Income)
            </span>
          </th>
          <th
            class=""
          >
            <span
              class=""
            >
              109.8% AMI (Area Median Income)
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class=""
          id="standardrow-0"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              1 person
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $51,300 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $61,300 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $71,300 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-1"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              2 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $58,600 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $68,600 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $78,600 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-2"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              3 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $65,950 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $75,950 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $75,950 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-3"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              4 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $73,250 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $83,250 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $83,250 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-4"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              5 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $79,100 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $99,100 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $140,000 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-5"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              6 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $85,000 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $95,000 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $150,000 per year
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <button
    aria-expanded="false"
    aria-label="Show more about maximum income based on household size"
    class="button has-icon-right is-inline font-medium md:bg-primary-lighter mt-3 text-primary-dark"
    tabindex="0"
  >
    <span
      class="button__content"
    >
      Show more
    </span>
    <span
      class="ui-icon ui-small button__icon "
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-angle-down "
        data-icon="angle-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M201.4 342.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 274.7 86.6 137.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
          fill="currentColor"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;

exports[`ListingDetailsHMITable renders ListingDetailsHMITable component with spinner before api call 1`] = `
<DocumentFragment>
  <div
    class="flex w-full justify-center"
  >
    <span
      class="ui-icon ui-large spinner-animation"
    >
      <svg
        fill="currentColor"
        viewBox="0 0 100 100"
      >
        <circle
          cx="50"
          cy="50"
          fill="none"
          r="32"
          stroke="currentColor"
          stroke-dasharray="150.79644737231007 52.26548245743669"
          stroke-width="9"
        />
      </svg>
    </span>
  </div>
</DocumentFragment>
`;

exports[`ListingDetailsHMITable renders ListingDetailsPricingTable component with open sale listing 1`] = `
<DocumentFragment>
  <div
    style="overflow-x: auto;"
  >
    <table
      class="w-full text-xs"
    >
      <thead>
        <tr>
          <th
            class=""
          >
            <span
              class=""
            >
              household size
            </span>
          </th>
          <th
            class=""
          >
            <span
              class=""
            >
              55% AMI (Area Median Income)
            </span>
          </th>
          <th
            class=""
          >
            <span
              class=""
            >
              82% AMI (Area Median Income)
            </span>
          </th>
          <th
            class=""
          >
            <span
              class=""
            >
              109.8% AMI (Area Median Income)
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class=""
          id="standardrow-0"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              1 person
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $51,300 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $61,300 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $71,300 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-1"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              2 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $58,600 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $68,600 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $78,600 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-2"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              3 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $65,950 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $75,950 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $75,950 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-3"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              4 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $73,250 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $83,250 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $83,250 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-4"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              5 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $79,100 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $99,100 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $140,000 per year
          </td>
        </tr>
        <tr
          class=""
          id="standardrow-5"
        >
          <td
            class="p-5"
            data-label="household size"
          >
            <span
              class="font-semibold"
            >
              6 people
            </span>
          </td>
          <td
            class="p-5"
            data-label="55% AMI (Area Median Income)"
          >
            $85,000 per year
          </td>
          <td
            class="p-5"
            data-label="82% AMI (Area Median Income)"
          >
            $95,000 per year
          </td>
          <td
            class="p-5"
            data-label="109.8% AMI (Area Median Income)"
          >
            $150,000 per year
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <button
    aria-expanded="false"
    aria-label="Show more about maximum income based on household size"
    class="button has-icon-right is-inline font-medium md:bg-primary-lighter mt-3 text-primary-dark"
    tabindex="0"
  >
    <span
      class="button__content"
    >
      Show more
    </span>
    <span
      class="ui-icon ui-small button__icon "
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-angle-down "
        data-icon="angle-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M201.4 342.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 274.7 86.6 137.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
          fill="currentColor"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;
